From ed74bb796eb3cd9885ef49caa8f7dde214aacb6c Mon Sep 17 00:00:00 2001
From: Stefan Lippers-Hollmann <s.l-h@gmx.de>
Date: Fri, 27 Nov 2009 13:32:00 +0000
Subject: [PATCH 5953/5953] rt2860/2870/3070/3090: use the firmware loader interface and library CRC code

Based on work by:
	Darren Salt <linux@youmustbejoking.demon.co.uk>
	Ben Hutchings <ben@decadent.org.uk>
---
 drivers/staging/rt2860/Kconfig                |    2 +
 drivers/staging/rt2860/common/firmware.h      |  558 -------------------------
 drivers/staging/rt2860/common/firmware_3070.h |  517 -----------------------
 drivers/staging/rt2860/common/rtmp_mcu.c      |   92 +++--
 drivers/staging/rt2870/Kconfig                |    2 +
 drivers/staging/rt3070/firmware.h             |  558 -------------------------
 drivers/staging/rt3070/md4.h                  |   42 --
 drivers/staging/rt3090/firmware.h             |  517 -----------------------
 8 files changed, 66 insertions(+), 2222 deletions(-)
 delete mode 100644 drivers/staging/rt2860/common/firmware.h
 delete mode 100644 drivers/staging/rt2860/common/firmware_3070.h
 delete mode 100644 drivers/staging/rt3070/firmware.h
 delete mode 100644 drivers/staging/rt3070/md4.h
 delete mode 100644 drivers/staging/rt3090/firmware.h

--- a/drivers/staging/rt2860/Kconfig
+++ b/drivers/staging/rt2860/Kconfig
@@ -1,7 +1,8 @@
 config RT2860
 	tristate "Ralink 2860/3090 wireless support"
-	depends on BROKEN
 	depends on PCI && X86 && WLAN
+	select CRC_CCITT
+	select FW_LOADER
 	select WIRELESS_EXT
 	select WEXT_PRIV
 	---help---
--- a/drivers/staging/rt2860/common/rtmp_mcu.c
+++ b/drivers/staging/rt2860/common/rtmp_mcu.c
@@ -37,27 +37,22 @@
 
 #include	"../rt_config.h"
 
-#if defined(RT2860) || defined(RT3090)
-#include "firmware.h"
-#include "../../rt3090/firmware.h"
-#endif
-#ifdef RT2870
-#include "../../rt3070/firmware.h"
-#include "firmware_3070.h"
-#endif
-
-#include <linux/bitrev.h>
+#include <linux/firmware.h>
+#include <linux/crc-ccitt.h>
 
 #ifdef RTMP_MAC_USB
 /* */
 /* RT2870 Firmware Spec only used 1 oct for version expression */
 /* */
 #define FIRMWARE_MINOR_VERSION	7
+#define FW_FILENAME_V1 "rt2870.bin"
+#define FW_FILENAME_V2 "rt3071.bin"
+MODULE_FIRMWARE(FW_FILENAME_V1);
+MODULE_FIRMWARE(FW_FILENAME_V2);
 #endif /* RTMP_MAC_USB // */
 
 /* New 8k byte firmware size for RT3071/RT3072 */
 #define FIRMWAREIMAGE_MAX_LENGTH	0x2000
-#define FIRMWAREIMAGE_LENGTH			(sizeof (FirmwareImage) / sizeof(u8))
 #define FIRMWARE_MAJOR_VERSION		0
 
 #define FIRMWAREIMAGEV1_LENGTH		0x1000
@@ -65,6 +60,10 @@
 
 #ifdef RTMP_MAC_PCI
 #define FIRMWARE_MINOR_VERSION		2
+#define FW_FILENAME_V1 "rt2860.bin"
+#define FW_FILENAME_V2 "rt3090.bin"
+MODULE_FIRMWARE(FW_FILENAME_V1);
+MODULE_FIRMWARE(FW_FILENAME_V2);
 #endif /* RTMP_MAC_PCI // */
 
 /*
@@ -111,44 +110,72 @@ int RtmpAsicLoadFirmware(struct rt_rtmp_
 {
 
 	int Status = NDIS_STATUS_SUCCESS;
-	u8 *pFirmwareImage = NULL;
 	unsigned long FileLength, Index;
 	u32 MacReg = 0;
 #ifdef RTMP_MAC_USB
 	u32 Version = (pAd->MACVersion >> 16);
 #endif
+	struct device		*dev;
+	const struct firmware	*fw = NULL;
+	const char		*name;
+	int			err;
 
 	/* New 8k byte firmware size for RT3071/RT3072 */
 	{
 #ifdef RTMP_MAC_PCI
+		dev = &((struct os_cookie *)pAd->OS_Cookie)->pci_dev->dev;
 		if (IS_RT3090(pAd) || IS_RT3390(pAd)) {
-			pFirmwareImage = FirmwareImage_3090;
-			FileLength = FIRMWAREIMAGE_MAX_LENGTH;
+			name = FW_FILENAME_V2;
+			FileLength = MAX_FIRMWARE_IMAGE_SIZE;
 		} else {
-			pFirmwareImage = FirmwareImage_2860;
-			FileLength = FIRMWAREIMAGE_MAX_LENGTH;
+			name = FW_FILENAME_V1;
+			FileLength = MAX_FIRMWARE_IMAGE_SIZE;
 		}
 #endif /* RTMP_MAC_PCI // */
 #ifdef RTMP_MAC_USB
-		/* the firmware image consists of two parts */
-		if ((Version != 0x2860) && (Version != 0x2872) && (Version != 0x3070)) {	/* use the second part */
-			/*printk("KH:Use New Version,part2\n"); */
-			pFirmwareImage =
-			    (u8 *)&
-			    FirmwareImage_3070[FIRMWAREIMAGEV1_LENGTH];
-			FileLength = FIRMWAREIMAGEV2_LENGTH;
-		} else {
-			/*printk("KH:Use New Version,part1\n"); */
-			if (Version == 0x3070)
-				pFirmwareImage = FirmwareImage_3070;
-			else
-				pFirmwareImage = FirmwareImage_2870;
+		dev = &((struct os_cookie *)pAd->OS_Cookie)->pUsb_Dev->dev;
+		if (Version == 0x2860 || Version == 0x2872 || Version == 0x3070) {
+			name = FW_FILENAME_V1;
 			FileLength = FIRMWAREIMAGEV1_LENGTH;
+		} else {
+			name = FW_FILENAME_V2;
+			FileLength = FIRMWAREIMAGEV2_LENGTH;
 		}
+		
 #endif /* RTMP_MAC_USB // */
+	err = request_firmware(&fw, name, dev);
+	if (err) {
+		dev_err(dev, "firmware file %s request failed (%d)\n",
+		        name, err);
+		goto fail;
 	}
 
-	RTMP_WRITE_FIRMWARE(pAd, pFirmwareImage, FileLength);
+	if (fw->size != FileLength) {
+		dev_err(dev, "firmware file %s size is wrong\n", name);
+		goto fail;
+	}
+
+	/* is it new enough? */
+	pAd->FirmwareVersion = (fw->data[fw->size - 4] << 8 |
+	                        fw->data[fw->size - 3]);
+
+	if (pAd->FirmwareVersion <
+	    (FIRMWARE_MAJOR_VERSION << 8 | FIRMWARE_MINOR_VERSION)) {
+		dev_err(dev, "rt28x0sta: firmware file %s is too old\n",
+		        name);
+		goto fail;
+	}
+
+	/* is the internal CRC correct? */
+	if (crc_ccitt(0xffff, fw->data, fw->size - 2) !=
+	    (fw->data[fw->size - 2] | (fw->data[fw->size - 1] << 8))) {
+		dev_err(dev, "firmware file %s failed internal CRC\n", name);
+		goto fail;
+	}
+
+	RTMP_WRITE_FIRMWARE(pAd, fw->data, fw->size);
+
+	release_firmware(fw);
 
 	/* check if MCU is ready */
 	Index = 0;
@@ -167,9 +194,14 @@ int RtmpAsicLoadFirmware(struct rt_rtmp_
 		Status = NDIS_STATUS_FAILURE;
 	}
 
+out:
 	DBGPRINT(RT_DEBUG_TRACE, ("<=== %s (status=%d)\n", __func__, Status));
 
 	return Status;
+fail:
+	release_firmware(fw);
+	Status = NDIS_STATUS_FAILURE;
+	goto out;
 }
 
 int RtmpAsicSendCommandToMcu(struct rt_rtmp_adapter *pAd,
--- a/drivers/staging/rt2870/Kconfig
+++ b/drivers/staging/rt2870/Kconfig
@@ -1,7 +1,8 @@
 config RT2870
 	tristate "Ralink 2870/3070 wireless support"
-	depends on BROKEN
 	depends on USB && X86 && WLAN
+	select CRC_CCITT
+	select FW_LOADER
 	select WIRELESS_EXT
 	select WEXT_PRIV
 	---help---
